lib/rubygems/resolver/activation_request.rb

```diff
@@ -49,15 +49,28 @@ def development?
   # Downloads a gem at +path+ and returns the file path.
 
   def download path
-    if @spec.respond_to? :source
+    Gem.ensure_gem_subdirectories path
+
+    if @spec.respond_to? :sources
+      exception = nil
+      path = nil
+      @spec.sources.each{ |source|
+        begin
+          path = source.download full_spec, path
+        rescue exception
+          nil
+        end
+      }
```

Comment:
the rescue is there to allow processing all sources and raise error only if none yields a result, pushed update here: https://github.com/stackbuilders/rubygems/blob/bfc57cad996452526c8093732ec66967bfc6ab60/lib/rubygems/resolver/activation_request.rb#L55-L63